File: 백준/Silver/10157. 자리배정/자리배정.py
import sys
input = sys.stdin.readline

# 입력 받기
C, R = map(int, input().split())
K = int(input())

# K가 전체 좌석 수보다 크면 배정 불가
if K > C * R:
    print(0)
    sys.exit()

# 반시계 방향 순서: 아래, 오른쪽, 위, 왼쪽
dx = [0, 1, 0, -1]
dy = [-1, 0, 1, 0]

# 공연장 좌석 초기화
seating = [[0] * C for _ in range(R)]
x, y = 0, R - 1  # 시작 위치 (1, 1) == (0, R - 1)
direction = 0    # 시작 방향: 아래

# K번째 좌석까지 배정
for num in range(1, K + 1):
    seating[y][x] = num  # 현재 위치에 대기 번호 배정
    
    # K번째 좌석에 도달하면 출력하고 종료
    if num == K:
        print(x + 1, R - y)
        break

    # 다음 위치 계산
    nx, ny = x + dx[direction], y + dy[direction]

    # 경계 체크 및 방문 확인
    if 0 <= nx < C and 0 <= ny < R and seating[ny][nx] == 0:
        x, y = nx, ny
    else:
        # 이동 불가하면 방향 전환
        direction = (direction + 1) % 4
        x, y = x + dx[direction], y + dy[direction]

File: 백준/Silver/10157. 자리배정/자리배정.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int C = Integer.parseInt(st.nextToken());
        int R = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(br.readLine());

        if (K > C * R) {
            System.out.println(0);
            return;
        }

        int[] dx = {0, 1, 0, -1};
        int[] dy = {-1, 0, 1, 0};

        int[][] seating = new int[R][C];
        int x = 0, y = R - 1;
        int direction = 0;

        for (int num = 1; num <= K; num++) {
            seating[y][x] = num;

            if (num == K) {
                System.out.println((x + 1) + " " + (R - y));
                break;
            }

            int nx = x + dx[direction];
            int ny = y + dy[direction];

            if (nx >= 0 && nx < C && ny >= 0 && ny < R && seating[ny][nx] == 0) {
                x = nx;
                y = ny;
            } else {
                direction = (direction + 1) % 4;
                x = x + dx[direction];
                y = y + dy[direction];
            }
        }
    }
}